### ISSUE TITLE
new verifier implemented in Rust

### ISSUE THREAD

**Issue by zookozcash** [zcash/zcash#2266]
Write a new verifier for the Zcash v1 (“Sapling”) ZK-proofs from the ground up in Rust. Run both the current verifier unchanged and the new verifier _in parallel_ so that it doesn't take anymore time (typically), even though it takes more CPU. Wait til both verifiers have completed. If they both return `true`, then the combo returns `true`. If they both return `false` then the combo returns `false`. If they return different things than each other, then the combo logs a detailed error message (including the proof in question and the current message and so on) and then aborts the process.

(This aborting-the-process part raises questions about blockchain forks and forward progress, but I'm pretty sure it is far more important to alert users and developers and to prevent users from accidentally accepting invalid payments than it is to minimize chances of a blockchain fork. Also, as the Zcash employees well learned during the recent security incident (https://z.cash/blog/security-announcement-2017-04-13.html), I'm firmly of the opinion that the _only_ way to communicate with most users is to abort the process.)

**Comment by daira** [comment]
What's the precise motivation for having two verifiers? Note that Sapling is a pefect opportunity to ditch the C++ prover and verifier *without* risking a consensus inconsistency: we would switch straight from using the old circuit in C++ to the new circuit in Rust, so there would not be a risk of inconsistency between C++ and Rust implementations. (Of course the Rust implementation would need to be thoroughly tested on a private network.)

**Comment by radix42** [comment]
Note that Mac and Windows downstream are currently only buildable with the --without-rust flag. Mac should be easy to add, windows, maybe not so much.

**Comment by zookozcash** [comment]
> What's the precise motivation for having two verifiers?

So that a bug in one implementation wouldn't allow counterfeiting. Both implementations would have to be buggy, and they would have to have the _same_ bug, or at least both of the bugs would have to be exploited by the same proof. ☺

> Note that Sapling is a perfect opportunity to ditch the C++ prover and verifier without risking a consensus inconsistency

We could do this before Sapling, or we could do it after Sapling. Realistically I guess we're probably not going to prioritise this before Sapling because [The Sapling Roadmap](https://z.cash/blog/the-near-future-of-zcash.html) is our priorities.

**Comment by daira** [comment]
The reason why I asked for the precise motivation, is that we may be able to achieve that motivation with less vulnerability to denial of service, or less technical debt.

In order to ensure that a bug in one implementation would not allow counterfeiting, it is sufficient to treat the transaction as invalid whenever either implementation says it is.

Reporting that the implementations are inconsistent is a separable problem. I know that you believe that this can only be reliably done by aborting, but we have not explored alternative approaches besides logging (which clearly is ineffective) and aborting. For example, going into RPC safe mode is another possibility that would also achieve the goal of protecting users, and would still be noticeable, but would be less likely to cause network partitions or an exploitable drop in network hash rate. I will file a separate ticket for discussion of how to report serious errors in a way that reliably attracts node operators' attention.

The reason I suggested switching {prover, verifier, circuit, curve} implementations *at* Sapling, rather than before or after, is that (at the cost of relying on only one implementation for each circuit) it avoids *ever needing* any of the following:

* implementation + review of the Groth16 proving system in C++
* implementation + review of the BLS12-381 curve arithmetic in C++
* implementation + review of the new circuit(s) in C++ using the libsnark API
* optimization + review of Bellman's existing PHGR13 proving system implementation
* optimization + review of Bellman's existing BN128 curve arithmetic implementation
* implementation + review of the compatibility circuit(s) using the Bellman API.

That's a *huge* amount of work that we can just skip, if we don't have redundant implementations. So the argument that "we have higher priorities" is invalid, because taking the approach I suggest involves *much* less work overall, not more. Indeed, avoiding the need for the above engineering work may be the only way to avoid biting off more for Sapling than we can chew. Any time saved will be available for more in-depth analysis and testing of the single implementation of each circuit.

**Comment by daira** [comment]
@zookozcash wrote:
> Run both the current verifier unchanged and the new verifier in parallel so that it doesn't take anymore time (typically), even though it takes more CPU.

This is not correct, because the verifier is already fully parallelised (I think). So any additional verification would take more time.

**Comment by ebfull** [comment]
I'm tracing the following thought process:

1. We don't like that the existing verifier is written in C++. We want it written in Rust.
2. If we already have a C++ verifier that works, we might as well run both of them if performance allows.

This seems sensible.

If it were the opposite and we only had a Rust-language implementation of the existing verifier, I wouldn't be as enthusiastic about building, auditing, maintaining and running a C++ verifier. Perhaps we can split that into a separate ticket.

**Comment by radix42** [comment]
My personal concern is getting rust tooling working right in my cross-compiled Windows build environment. I haven't gotten that setup yet. And it may or may not be possible:
https://github.com/japaric/rust-cross#i-want-to-build-binaries-for-linux-mac-and-windows-how-do-i-cross-compile-from-linux-to-mac

**Comment by ebfull** [comment]
Cross compilation with Rust is definitely possible; I was doing it two or three years ago from Linux to Windows. I'm not sure about Mac but I don't know why it would be much different. Note that the guide you're reading was posted a while ago, since then the tooling (i.e. rustup) has been made more amicable to cross compilation and building standard libraries for other targets.

**Comment by radix42** [comment]
oh that's excellent, I admit that I did not pay attention to when that was posted.

**Comment by daira** [comment]
@ebfull wrote:
> I'm tracing the following thought process:
>
> 1. We don't like that the existing verifier is written in C++. We want it written in Rust.
> 2. If we already have a C++ verifier that works, we might as well run both of them if performance allows.
>
>This seems sensible.
>
> If it were the opposite and we only had a Rust-language implementation of the existing verifier, I wouldn't be as enthusiastic about building, auditing, maintaining and running a C++ verifier. Perhaps we can split that into a separate ticket.

There's no "might as well". We don't have a Rust implementation of the Sprout circuit, nor do we have a C++ implementation of BLS12-381, Groth16, or the Sapling circuit. We have to assess the engineering costs of writing those, and the increased performance cost of verification, against the advantage of being able to detect inconsistent verification results between redundant implementations.

Note that whether to do redundant verification for the old circuit is a separable decision to whether to do it for the new circuit. Also, we have the option of switching to a Rust implementation of verification for the old circuit *at our leisure* after the Sapling hard fork &mdash; and then it only has to work on the transactions that were actually in the pre-fork chain. So we will not have the technical debt of maintaining the C++ verifier indefinitely, regardless of whether we do redundant verification.

(This doesn't preclude us writing a Rust implementation of the Sprout circuit and running it off-line, or on a detector node, in order to increase confidence that the C++ circuit and verifier was correct. However, it takes doing that out of the critical path for Sapling.)

**Comment by daira** [comment]
Arguing against my previous comment in part:

Writing a Rust implementation of the Sprout circuit is potentially useful experience. It is not the case that the Sprout and Sapling circuits are completely independent; Sapling is a refinement of Sprout, and we can reuse code between them, even if it is only by copying and incremental modification rather than direct code sharing. The process of writing and testing the Sprout circuit in Rust could reveal security bugs in the C++ circuit implementation or in libsnark, in time to usefully mitigate them before the Sapling hard fork. It would increase confidence that the Rust APIs for expressing circuits are complete. This also does not leave us with any significant technical debt: after the Sapling hard fork, we just turn off the C++ verifier for pre-fork blocks, and then are left with a clean Rust code base for the zero-knowledge layer, that is still able to verify the whole historical block chain.

I am much less convinced that we should do redundant verification for the Sapling circuit. First of all, it's more work to write that, because we are writing in a more error-prone language, and because we would need to implement BLS12-381 and Groth16 in libsnark (whereas we already have an implementation of [BN-254 in Rust](https://github.com/zcash/bn), and I think also PHGR13?). Second, if we launched the Sapling fork with redundant verification, we'd then be afraid to turn off the C++ verifier in case that introduced security vulnerabilities. That would indefinitely leave us with technical debt, attack surface due to a non-memory-safe language processing untrusted transaction input, and the ongoing performance cost of the redundant verification.

**Comment by ebfull** [comment]
Note that we don't really need to implement the circuit in Rust in order to verify Sprout zk-SNARKs.

**Comment by daira** [comment]
Oh, good point @ebfull. And similarly, we don't need to implement the Sapling circuit in C++ to do verification of Sapling proofs via libsnark; we would only need to implement BLS12-381 and Groth16 (including deserializing a Groth16 verification key) in libsnark.

**Comment by jasondavies** [comment]
With @ebfull's help, I've written a [pure Rust verifier](https://github.com/plutomonkey/zcash-sprout-verifier) that uses [bn](https://github.com/zcash/bn) to verify the zk-SNARK proofs used in Zcash, and I've now independently verified all proofs to date on mainnet and testnet, thereby increasing confidence that no-one has yet found and exploited a bug in libsnark's verifier.

Currently, the code is as straightforward as possible and doesn't use any fancy optimisations, instead opting to mirror the procedure in the paper as closely as possible.  The verification code itself is [quite short](https://github.com/plutomonkey/zcash-sprout-verifier/blob/master/src/lib.rs#L29-L49).

My [rust-verifier branch](https://github.com/plutomonkey/zcash/tree/rust-verifier) of zcash can be used to test the verifier.

**Comment by daira** [comment]
What we ended up doing was to:
* write an implementation of the Sprout circuit in Rust to the bellman API, and use it from Sapling activation onward;
* switch to BLS12-381 and Groth16 for Sprout proofs from Sapling activation onward (so, no implementation of BN-254 or PHGR13 in Rust was needed);
* not implement anything else associated with proving or verifying (other than just using librustzcash) in C++.
